add to MSAL contributing guide (#4388)
AzureAD/microsoft-authentication-library-for-dotnet: CONTRIBUTING.md

Please visit [https://cla.microsoft.com/](https://cla.microsoft.com/) and sign the Contributor License Agreements.  You only need to do that once. We can not look at your code until you've submitted this request.

> ## Contributing your own code changes

Reviewer: @localden Thoughts on moving the section on Submitting a bug/feature request above this one?

If you would like to work on an involved feature, please file a design proposal first; more instructions can be found below, under [Before writing code](#before-writing-code).

### Before writing code

We've seen pull requests, where customers would solve an issue in a way which either wouldn't fit into the end-to-end design because of how it's implemented, or it would affect the design in a way, which is not something we'd like to do. To avoid these situations and potentially save you a lot of time, we encourage customers to discuss the preferred design with the team first. To do so, file a [new design proposal issue]((https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/new?assignees=&labels=design-proposal&projects=&template=design_proposal.md)), link it to the issue you'd like to address, and provide detailed information about how you'd like to solve a specific problem.

### Identifying scale

> If you would like to contribute to one of our repositories, first identify the scale of what you would like to change. If it is small (grammar/spelling or a bug fix), feel free to start working on a fix. If you are submitting a feature or a substantial code contribution, please discuss it with the team and ensure it follows the product roadmap. You might also read these two blogs posts on contributing code: [Open Source Contribution Etiquette by Miguel de Icaza](http://tirania.org/blog/archive/2010/Dec-31.html) and [Don't "Push" Your Pull Requests by Ilya Grigorik](https://www.igvita.com/2011/12/19/dont-push-your-pull-requests/). All code submissions will be rigorously reviewed and tested further by the team, and only those that meet an extremely high bar for both quality and design/roadmap appropriateness will be merged into the source.

Reviewer: Good articles!

Reviewer 2: +1 , nice articles

- You posted a high-level description of how the changes will be implemented and received a positive acknowledgement from the team before getting too committed to the approach or investing too much effort in implementing it.
- You added test coverage following existing patterns within the codebase.
- Your code matches the existing syntax conventions within the codebase.
- Your PR is small, focused, and avoids making unrelated changes.

If your pull request contains any of the below, it's less likely to be merged.

- Changes that break backward compatibility
- Changes that are only wanted by one person or company. Changes need to benefit a large enough proportion of developers using our auth libraries.
- Changes that add entirely new feature areas without prior agreement
- Changes that are mostly about refactoring existing code or code style

Very large PRs that would take hours to review (remember, we're trying to help lots of people at once). For larger work areas, please discuss with us to find ways of breaking it down into smaller, incremental pieces that can go into separate PRs.

### During a pull request review

> A core contributor will review your pull request and provide feedback. To ensure that there is not a large backlog of inactive PRs, the pull request will be marked as stale after two weeks of no activity. After another four days, it will be closed.

Reviewer: Has this automation been completed in other repos?

Author: No, this has not and I don't think we plan to.

Reviewer: I've merged this into the Merging a pull request section.

If you're not sure how to create a pull request, read GitHub's [About pull requests
](https://help.github.com/articles/using-pull-requests) article. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. The coding style and general engineering guidelines are published on the Engineering guidelines page.

### Ensuring test coverage

- Tests need to be provided for every bug and feature that is completed.
  - Unit tests must cover all new aspects of the code.
- Before and after performance and stress tests results are successfully evaluated - no regressions are allowed.
- Performance and stress tests are extended as relevant.

Note that you won't be able to run integration tests locally because they connect to Azure Key Vault to fetch some test users and passwords.

### Pull request feedback

Your pull request will now go through extensive checks by the subject matter experts on our team. Please be patient; we have hundreds of pull requests across all of our repositories. Update your pull request according to feedback until it is approved by one of the team members.

### Merging a pull request

> When your pull request has had all feedback addressed, it has been signed off by one or more reviewers with commit access, and all checks are green, we will commit it.

Reviewer: This is a bit inaccurate since we don't run CI on forks. Idk if we need to rephrase.

Author: @pmaytak I didn't know that! 😁 How would you like to see it rephrased?

Reviewer: It actually applies to all of our GitHub repos. I've reworded this section.

- You add test coverage following existing patterns within the codebase
- Your code matches the existing syntax conventions within the codebase
- Your PR is small, focused, and avoids making unrelated changes

If your pull request contains any of the below, it's less likely to be merged.

- Changes that break backward compatibility
- Changes that are only wanted by one person/company. Changes need to benefit a large enough proportion of developers using our auth libraries.
- Changes that add entirely new feature areas without prior agreement
- Changes that are mostly about refactoring existing code or code style

Very large PRs that would take hours to review (remember, we're trying to help lots of people at once). For larger work areas, please discuss with us to find ways of breaking it down into smaller, incremental pieces that can go into separate PRs.

## During pull request review

A core contributor will review your pull request and provide feedback. To ensure that there is not a large backlog of inactive PRs, the pull request will be marked as stale after two weeks of no activity. After another four days, it will be closed.

## Submitting a pull request

> If you're not sure how to create a pull request, read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. The coding, style, and general engineering guidelines are published on the Engineering guidelines page.

Reviewer: ```suggestion
If you're not sure how to create a pull request, read GitHub's [About pull requests
](https://help.github.com/articles/using-pull-requests) article. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. The coding style and general engineering guidelines are published on the Engineering guidelines page.
```

> ## Writing Tests

Reviewer: ```suggestion
## Ensuring test coverage
```

> - Tests need to be provided for every bug/feature that is completed.
> - Tests need to be provided for every bug/feature that is completed.
>   - Unit tests cover all new aspects of the code.
> - Before and after performance and stress tests results are evaluated (no regressions allowed).

Reviewer: ```suggestion
- Tests need to be provided for every bug and feature that is completed.
  - Unit tests must cover all new aspects of the code.
- Before and after performance and stress tests results are successfully evaluated - no regressions are allowed.
```

- Performance and stress tests are extended as relevant.

> ## Feedback

Reviewer: ```suggestion
## Pull request feedback
```

Your pull request will now go through extensive checks by the subject matter experts on our team. Please be patient; we have hundreds of pull requests across all of our repositories. Update your pull request according to feedback until it is approved by one of the team members.

> ## Merging pull requests

Reviewer: ```suggestion
## Merging a pull request
```
